Bitbucket cloud authentication

FilipSwiatczak opened the pull request:
## Hey, I just made a Pull Request!

This introduces Bitbucket Auth option for login. Based on existing auth clients, with necessary updates. 
Auth handler and sign-in-resolver as per #6245 

There is no default Resolver, please choose one of two in `plugins/auth-backend/src/providers/factories.ts`

```
export const factories: { [providerId: string]: AuthProviderFactory } = {
  ...
  bitbucket: createBitbucketProvider({
    //  or   bitbucketUserIdSignInResolver
    signIn: { resolver: bitbucketUsernameSignInResolver },
  }),
};
```
To match against your catalog user entity, annotation will be required depending on above choice and should be equal to bitbucket profile value:
```
annotations:
    bitbucket.org/user-id: "{80dc88eXXXXXXcc5265}"
OR
annotations:
    bitbucket.org/username: filip_swiatczak
```

[Bitbucket user 2.0 API](https://developer.atlassian.com/bitbucket/api/2/reference/resource/user)
![image](https://user-images.githubusercontent.com/30108240/134557265-0bf5bc1e-c4ce-494d-8c51-3ead9a33e9c6.png)
![image](https://user-images.githubusercontent.com/30108240/135340778-e56a3377-3d2d-4411-ba9e-5950d8393604.png)
![image](https://user-images.githubusercontent.com/30108240/134556785-91be8dca-1a1a-4e1b-8295-cbc91ca55ee9.png)

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [X] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

regicsolutions commented:
Nice! Looks like this will unfortunately only support Bitbucket Cloud and not Bitbucket Server given it only supports OAuth 1.0a using application links http://www.passportjs.org/packages/passport-atlassian-oauth/

FilipSwiatczak commented:
> Nice! Looks like this will unfortunately only support Bitbucket Cloud and not Bitbucket Server given it only supports OAuth 1.0a using application links
> 
> http://www.passportjs.org/packages/passport-atlassian-oauth/
> 
> https://bitbucket.org/atlassianlabs/passport-atlassian-oauth/src/master/
hi @regicsolutions, yes, I've updated the PR title to reflect the cloud only aspect, for the moment anyway 
It utilises passport-bitbucket-oauth2 (http://www.passportjs.org/packages/passport-bitbucket-oauth2/) so OAuth 2.0 API
I have a bitbucket server instance as well so might be able to find a way to do both, but we are migrating away from server so wasn't a priority. Cheers

regicsolutions commented:
> > Nice! Looks like this will unfortunately only support Bitbucket Cloud and not Bitbucket Server given it only supports OAuth 1.0a using application links
> 
> > 
> 
> > http://www.passportjs.org/packages/passport-atlassian-oauth/
> 
> > 
> 
> > https://bitbucket.org/atlassianlabs/passport-atlassian-oauth/src/master/
> 
> 
> 
> hi @regicsolutions, yes, I've updated the PR title to reflect the cloud only aspect, for the moment anyway 
> 
> It utilises passport-bitbucket-oauth2 (http://www.passportjs.org/packages/passport-bitbucket-oauth2/) so OAuth 2.0 API
> 
> I have a bitbucket server instance as well so might be able to find a way to do both, but we are migrating away from server so wasn't a priority. Cheers

That would be great, Cheers!

Rugvip commented on docs/auth/bitbucket/provider.md:
```diff
@@ -0,0 +1,70 @@
+description: Adding Bitbucket OAuth as an authentication provider in Backstage
+---
+
+The Backstage `core-api` package comes with a Bitbucket authentication provider
```
```suggestion
The Backstage `core-pluign-api` package comes with a Bitbucket authentication provider
```

Rugvip commented on docs/auth/bitbucket/provider.md:
```diff
@@ -0,0 +1,70 @@
+  `b59241722e3c3b4816e2`
+- `clientSecret`: The Secret tied to the generated Key.
+
+## Adding the provider to the Backstage frontend
```
Let's just reference the parent doc like the other providers do in this section

FilipSwiatczak replied:
Done!

Rugvip commented on packages/core-app-api/src/apis/implementations/auth/bitbucket/BitbucketAuth.ts:
```diff
@@ -0,0 +1,142 @@
+  icon: BitbucketIcon,
+};
+
+class BitbucketAuth implements OAuthApi, SessionApi {
```
It should be possible to implement this like the other standard OAuth providers like e.g. the google one, i.e. we just need the static `create` method that in turn calls out to `OAuth2.create`

FilipSwiatczak replied:
I've followed the google example and refactored, thanks!

Rugvip commented on plugins/auth-backend/src/providers/bitbucket/provider.ts:
```diff
@@ -0,0 +1,124 @@
+  }
+}
+
+export type BitbucketProviderOptions = {};
```
Optional, but if you want to it'd be nice to implement sign-in resolver and auth handler from #6245 straight away

FilipSwiatczak replied:
Hi @Rugvip thank you for the pointers, this should be done now :)

Rugvip reviewed (commented):
Thank you! 👍 😁 

Looks good overall, there's some opportunity for cleanup but otherwise ready to :shipit:

FilipSwiatczak commented:
thanks @Rugvip I've addressed your suggestions and also added avatarUrl to profile object passed into authHandler - that way PassportStrategyHelper picks it up for the settings profile icon. 
What I noticed though is that catalog user entry, when matched, does not get the icon updated, minor thing but made me wonder. Let me know if anything else needs work! Cheers :)

FilipSwiatczak commented:
thanks @Rugvip I've addressed your suggestions and also added avatarUrl to profile object passed into authHandler - that way PassportStrategyHelper picks it up for the settings profile icon. 
What I noticed though is that catalog user entry, when matched, does not get the icon updated, minor thing but made me wonder. Let me know if anything else needs work! Cheers :)

Rugvip commented on plugins/auth-backend/src/providers/bitbucket/provider.ts:
```diff
@@ -0,0 +1,313 @@
+  };
+};
+
+export const createBitbucketProvider = (
```
Need to make sure this and the mail resolver is exported, or it won't be possible to configure a custom provider

FilipSwiatczak replied:
createBitbucketProvider  and bitbucketEmailSignInResolver are both exported, see below IDE is happy in factories.ts
bitbucketDefaultSignInResolver is not exported, should be all good

![image](https://user-images.githubusercontent.com/30108240/135335350-02015566-e7fe-47a5-9560-6df329e1136a.png)

Rugvip replied:
Exported from the `@backstage/auth-backend` package to be precise. I'm looking for something new to show up in the API report :grin:

Rugvip replied:
Bump to make sure this doesn't fall between the cracks :grin:

FilipSwiatczak replied:
![image](https://user-images.githubusercontent.com/30108240/135416389-28872271-74cd-4c66-8fd2-817ddba1b5d4.png)
exported and inside api-report.md for @backstage/auth-backend, all good I hope?

Rugvip reviewed (commented):
Thank you for refactoring to the new sign-in resolver pattern! I realized one small cleanup we can do while we're at it, but otherwise I'd say we're good to go :shipit:

FilipSwiatczak commented on plugins/auth-backend/src/providers/bitbucket/provider.ts:
```diff
@@ -0,0 +1,313 @@
+
+  const entity = await ctx.catalogIdentityClient.findUser({
+    annotations: {
+      'bitbucket/email': profile.email,
```
yup yup

FilipSwiatczak commented on plugins/auth-backend/src/providers/bitbucket/provider.ts:
```diff
@@ -0,0 +1,313 @@
+    /**
+     * Maps an auth result to a Backstage identity for the user.
+     */
+    resolver?: SignInResolver<OAuthResult>;
```
done

FilipSwiatczak commented:
cheers @Rugvip, should be final go now fingers crossed. Updated from your comments + fixed an omission in provider test.

regicsolutions commented:
@FilipSwiatczak could we maybe keep it 'Sign into Bitbucket' vs 'Sign into Bitbucket Cloud?

Rugvip commented on plugins/auth-backend/src/providers/bitbucket/provider.ts:
```diff
@@ -0,0 +1,313 @@
+
+import express from 'express';
+import passport, { Profile as PassportProfile } from 'passport';
+// @ts-ignore passport-bitbucket-oauth2 does not have type definitions
```
Missed this one, let's use a module type definition instead, it can be very lightweight, see for example https://github.com/backstage/backstage/blob/1b7700a80a4a52a8149849d37cb1a5ab281e6373/plugins/auth-backend/src/providers/okta/types.d.ts#L16-L20

FilipSwiatczak replied:
Done

FilipSwiatczak commented:
thanks @Rugvip, I missed the index export propagation! Fixed now :) 'Sign in using Bitbucket' updated

Rugvip commented on plugins/auth-backend/src/providers/bitbucket/provider.ts:
```diff
@@ -0,0 +1,313 @@
+          });
+
+      const signInResolverFn =
+        options?.signIn?.resolver ?? bitbucketDefaultSignInResolver;
```
I think this fell between the cracks a bit too. Let's completely remove the `bitbucketDefaultSignInResolver` and not have any default resolver

FilipSwiatczak replied:
I've done the types.d.ts, before I push I'll add this as well. Sorry I don't see how removing the default resolver works. Above quoted line is the precise mechanism allowing user to plug in the Email-resolver aka. `'bitbucket.org/email': profile.email` annotation. 
The default at the moment uses : `fullProfile.username || fullProfile.id;` which I think would be preferable to most? 
By removing do you mean restructuring so that the username/id logic still remains the default?

Rugvip replied:
Removing it as the default would be to just change the line to `const signInResolverFn = options?.signIn?.resolver`

Part of the change we're making with the sign-in resolvers is to not have the auth providers act as sign-in providers by default. We've avoided that breaking change for now as we migrate existing providers by just having a default resolver that implements whatever the existing logic is. The default resolvers aren't supposed to stick around though, and once we migrated all auth providers we remove all defaults of them in a breaking change.

Good point about this default resolver being useful though. I think it could make sense to keep, but let's call it something that describes how it works, something like `bitbucketUserIdSignInResolver`? We'd also need to add the call out to the catalog and `getEntityClaims` to make sure it creates a proper identity that exists.

FilipSwiatczak replied:
@Rugvip cool I'll make bitbucketUserIdSignInResolver with call to catalog and getEntityClaims, do you think it's good idea to create a user in catalog if missing? Like new joiners first logging in first time? Or rely on other ingestion processor? Though I think the processor polling is being moved away from due to eating up rate limits. Cheers

Rugvip replied:
That's not possible at the moment unfortunately. The users need to already be present in the catalog, ingested via some other means

FilipSwiatczak replied:
3 resolvers are now added, and default removed
  bitbucketUsernameSignInResolver,
  bitbucketUuidSignInResolver,
  bitbucketDisplayNameSignInResolver,
annotations:
    bitbucket.org/uuid: "{80dc88eXXXXXXcc5265}"
OR
annotations:
    bitbucket.org/displayName: Filip Swiatczak
OR
annotations:
    bitbucket.org/username: filip_swiatczak
    
So user decides which one to use and creates corresponding annotations! I will add more info in the main description

FilipSwiatczak commented:
hi @Rugvip after reading into the class structure and types I realised catalogIdentityClient.findUser and UserQuery type focus on annotations and since that seems to be the focus (as opposed to reading metadata.name from catalog for example), the previous default resolver didn't have a good way to translate. 

Instead I focused on the Bitbucket /2.0/user API and it's response as available under result.fullProfile. Narrowing to values interesting as links to users. 
```
{
  "id": "{80dc88eXXXXXXX7cc5265}",
  "displayName": "Filip Swiatczak",
  "username": "filip_swiatczak",
}
```
So I simply made three Resolvers:
  ,
  ,
  bitbucketDisplayNameSignInResolver,
that are exported and can be used in factories.ts to pick whichever of above 3 fits best. User entities can therefore contain:

annotations:
   bitbucket.org/uuid: "{80dc88eXXXXXXcc5265}" for **bitbucketUuidSignInResolver** used
OR
annotations:
   bitbucket.org/displayName: Filip Swiatczak for **bitbucketDisplayNameSignInResolver** used
OR
annotations:
    bitbucket.org/username: filip_swiatczak  for **bitbucketUsernameSignInResolver** used

Rugvip commented:
@FilipSwiatczak Sounds good! I think perhaps drop the display name one since it's a bit of a security sensitive mapping we're doing here and I suspect that one might not have the guarantees we want. I think username makes sense as you suggest and perhaps we could just talk about "user ID" since UUIDs have a more specific meaning and `id` is what we get from Bitbucket? So prolly `bitbucket.org/user-id` and `bitbucketUserIdSignInResolver` as just "id" would be a bit too generic

FilipSwiatczak commented:
> @FilipSwiatczak Sounds good! I think perhaps drop the display name one since it's a bit of a security sensitive mapping we're doing here and I suspect that one might not have the guarantees we want. I think username makes sense as you suggest and perhaps we could just talk about "user ID" since UUIDs have a more specific meaning and `id` is what we get from Bitbucket? So prolly `bitbucket.org/user-id` and `bitbucketUserIdSignInResolver` as just "id" would be a bit too generic

Thanks @Rugvip that's done now!

Rugvip commented on plugins/auth-backend/src/providers/bitbucket/provider.ts:
```diff
@@ -0,0 +1,316 @@
+  /**
+   * Configure sign-in for this provider, without it the provider can not be used to sign users in.
+   */
+  signIn: {
```
```suggestion
  signIn?: {
```

One last thing 😅

FilipSwiatczak replied:
@Rugvip I thought the whole point of removing default is to make a choice of resolver mandatory, otherwise how would the catalog findUser know what annotation to look for?

Rugvip replied:
So without a sign-in resolver you wouldn't be able to use the provider to sign-in to Backstage at all. More concretely there won't be any `BackstageIdentity` in the response. The provider will still be usable to delegate access to resources, so you can still for example use it to get a Bitbucket token that you can then use to talk to the Bitbucket API.

FilipSwiatczak replied:
hi @Rugvip it doesn't work. I've applied above changes. When signInResolver is undefined, something is still trying to access backstageIdentity.Id from the return from handleResult() call.
With this set, it gets weird type error by the way:
```
      const signInResolver: SignInResolver<BitbucketOAuthResult> | undefined =
        options?.signIn?.resolver
          ? (info =>
            options?.signIn?.resolver(info, {
              catalogIdentityClient,
              tokenIssuer,
              logger,
            }))
          : undefined;
```
Above is checking and returning undefined if resolver option is left unset. 
I can verify that this.signInResolver is indeed undefined inside handleResult and so setting of backstage identify is skipped  and yet this is the outcome below.

I'll attempt again to fix it tomorrow but meanwhile if it's obvious to you please feel free to make that final change. Thank you for your help!
![image](https://user-images.githubusercontent.com/30108240/135915563-27f9dd8b-f6ea-4703-a1ee-241db5e7e39b.png)

FilipSwiatczak replied:
@Rugvip restoring minimal form of default resolver fixes it, as long as it returns {id: userId} example:
```
      const defaultResolver: SignInResolver<BitbucketOAuthResult> =
        async (info, ctx) => {
          const { fullProfile } = info.result;

          const userId = fullProfile.username || fullProfile.id;

          const token = await ctx.tokenIssuer.issueToken({
            claims: { sub: userId, ent: [`user:default/${userId}`] },
          });

          return { id: userId, token };
        };

      const signInResolverFn =
        options?.signIn?.resolver ?? defaultResolver;
```

This, as I understand, would still allow for:

> you can still for example use it to get a Bitbucket token that you can then use to talk to the Bitbucket API

and for users choosing Resolvers in factories.ts. 
Does that sound acceptable?

Rugvip commented on plugins/auth-backend/src/providers/bitbucket/provider.ts:
```diff
@@ -0,0 +1,316 @@
+              profile: makeProfileInfo(fullProfile, params.id_token),
+            });
+
+      const signInResolver: SignInResolver<BitbucketOAuthResult> = info =>
```
So if `options.signIn` isn't set, this should be undefined

Rugvip commented on plugins/auth-backend/src/providers/factories.ts:
```diff
@@ -39,4 +43,7 @@ export const factories: { [providerId: string]: AuthProviderFactory } = {
   onelogin: createOneLoginProvider(),
   awsalb: createAwsAlbProvider(),
+  bitbucket: createBitbucketProvider({
+    signIn: { resolver: bitbucketUsernameSignInResolver },
```
Along with making it optional we drop this here as well

Rugvip reviewed (approved):
Thank you for the rounds of iteration! 👍 😅 

I'm thinking we :shipit: as is and I'll have a look at what's up with the `backstageIdentity` being required. That seems like a bug that's unrelated to this code.

FilipSwiatczak commented:
> Thank you for the rounds of iteration! 👍 😅
> 
> I'm thinking we :shipit: as is and I'll have a look at what's up with the `backstageIdentity` being required. That seems like a bug that's unrelated to this code.

thank you for all your help @Rugvip! Much appreciated and excited to have made my first contribution 👍 Onto the next :smile:

regicsolutions commented:
@FilipSwiatczak It looks like Bitbucket Server just introduced OAuth 2.0 support in there latest [7.20 Release](https://confluence.atlassian.com/bitbucketserver/bitbucket-data-center-and-server-7-20-release-notes-1101934428.html) i just gave this a try and it seems like this provider was statically set to redirect to a bitbucket.org url, any suggestions on how we can maybe make this dynamic to support both Bitbucket Cloud and Serve?

`https://bitbucket.org/site/oauth2/authorize?response_type=code&redirect_uri=http ://localhost:7000/api/auth/bitbucket/handler/frame&scope=team&state=STATE&client_id=CLIENT_ID` 

According to this [document ](https://confluence.atlassian.com/bitbucketserver/bitbucket-oauth-2-0-provider-api-1108483661.html) looks like the bitbucket server url would be `https://bitbucketserver.myorg.com/rest/oauth2/latest/authorize?client_id=CLIENT_ID&redirect_uri=REDIRECT_URI&response_type=code&state=STATE&scope=SCOPE&code_challenge=CODE_CHALLENGE&code_challenge_method=S256`

regicsolutions commented:
> @FilipSwiatczak It looks like Bitbucket Server just introduced OAuth 2.0 support in there latest [7.20 Release](https://confluence.atlassian.com/bitbucketserver/bitbucket-data-center-and-server-7-20-release-notes-1101934428.html) i just gave this a try and it seems like this provider was statically set to redirect to a bitbucket.org url, any suggestions on how we can maybe make this dynamic to support both Bitbucket Cloud and Server 7.20?
> 
> 
> 
> `https://bitbucket.org/site/oauth2/authorize?response_type=code&redirect_uri=http ://localhost:7000/api/auth/bitbucket/handler/frame&scope=team&state=STATE&client_id=CLIENT_ID` 
> 
> 
> 
> According to this [document ](https://confluence.atlassian.com/bitbucketserver/bitbucket-oauth-2-0-provider-api-1108483661.html) looks like the bitbucket server url would be `https://bitbucketserver.myorg.com/rest/oauth2/latest/authorize?client_id=CLIENT_ID&redirect_uri=REDIRECT_URI&response_type=code&state=STATE&scope=SCOPE&code_challenge=CODE_CHALLENGE&code_challenge_method=S256`
> 
> 

Friendly ping @FilipSwiatczak 😀

p-kamath commented:
@regicsolutions - did you happen to get it working for bitbucketServer?
